h2o-core/src/main/java/hex/PartialDependence.java

```diff
@@ -94,151 +121,189 @@ private void checkSanityAndFillParams() {
-          if (actualbins == 1) delta = 0;
-          for (int j = 0; j < colVals.length; ++j) {
-            colVals[j] = v.min() + j * delta;
-          }
-        }
-
-        if (_add_missing_na)
-          colVals[actualbins] = Double.NaN; // set last bin to contain nan
-
-        Log.debug("Computing PartialDependence for column " + col + " at the following values: ");
-        Log.debug(Arrays.toString(colVals));
+        double[] colVals = extractColValues(col, _nbins, fr.vec(col));
+        double[] col2Vals = workingOn1D ? null : extractColValues(col2, _nbins, fr.vec(col2));
 
         Futures fs = new Futures();
-        final double meanResponse[] = new double[colVals.length];
-        final double stddevResponse[] = new double[colVals.length];
-        final double stdErrorOfTheMeanResponse[] = new double[colVals.length];
+        int responseLength = workingOn1D ? colVals.length : colVals.length * col2Vals.length;
+        final double meanResponse[] = new double[responseLength];
+        final double stddevResponse[] = new double[responseLength];
+        final double stdErrorOfTheMeanResponse[] = new double[responseLength];
 
         final boolean cat = fr.vec(col).isCategorical();
+        final boolean cat2 = workingOn1D?false:fr.vec(col2).isCategorical();
         // loop over column values (fill one PartialDependence)
         for (int k = 0; k < colVals.length; ++k) {
           final double value = colVals[k];
-          final int which = k;
-          H2O.H2OCountedCompleter pdp = new H2O.H2OCountedCompleter() {
-            @Override
-            public void compute2() {
-              Frame fr;
-              if (_row_index > 0) {
-                fr = Rapids.exec("(rows " + _frame_id + "  " + _row_index + ")").getFrame();
-              } else {
-                fr = _frame_id.get();
-              }
-              Frame test = new Frame(fr.names(), fr.vecs());
-              Vec orig = test.remove(col);
-              Vec cons = orig.makeCon(value);
-              if (cat) cons.setDomain(fr.vec(col).domain());
-              test.add(col, cons);
-              Frame preds = null;
-              try {
-                preds = _model_id.get().score(test, Key.make().toString(), _job, false);
-                if (preds == null || preds.numRows() == 0) {  // this can happen if algo will not predict on rows with NAs
-                  meanResponse[which] = Double.NaN;
-                  stddevResponse[which] = Double.NaN;;
-                  stdErrorOfTheMeanResponse[which] = Double.NaN;;
+
+          int secondColLen = workingOn1D?1:col2Vals.length;
+          for (int j = 0; j < secondColLen; j++) {  // will go through at least once
+            final double value2 = workingOn1D?Double.NaN:col2Vals[j];
+            final int which = workingOn1D?k:j+k*col2Vals.length;
+            final boolean workOn2D = !workingOn1D;
+            
+            H2O.H2OCountedCompleter pdp = new H2O.H2OCountedCompleter() {
```

Comment:
I think an anonymous inner class of this size deserves a name

Reply:
I agree.  However, it seems like this is a normal practice here at H2O.  At least with the people who worked here before.

Reply:
I agree with @honzasterba. Better readability this way.

Reply:
there are many things that are frequent across the h2o codebase, but really should not be.
let's improve!

Reply:
Be the change you want to see in this codebase @wendycwong ;)

Reply:
OMG, did you guys just finished watching some inspiring movies/shows or something?  Sure, I will take care of it.

Reply:
Done moving pdp to its own glorious class.